File: frontend/src/app/features/inventory/procurement/tax-rule/tax-rule.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../../environments';
import {
  Page,
  TaxRule,
  TaxRuleRequest,
  TaxRuleStatusUpdateRequest,
  TaxRuleStatusUpdateResponse,
} from './tax-rule.model';

@Injectable({ providedIn: 'root' })
export class TaxRuleService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/inventory/procurement/tax-rules`;

  getAll(activeOnly = false): Observable<TaxRule[]> {
    const params = new HttpParams().set('activeOnly', activeOnly);
    return this.http.get<TaxRule[]>(this.baseUrl, { params });
  }

  getPage(p: { search?: string; page?: number; size?: number }): Observable<Page<TaxRule>> {
    let params = new HttpParams().set('page', p.page ?? 0).set('size', p.size ?? 25);
    if (p.search) params = params.set('search', p.search);
    return this.http.get<Page<TaxRule>>(`${this.baseUrl}/page`, { params });
  }

  getById(id: number): Observable<TaxRule> {
    return this.http.get<TaxRule>(`${this.baseUrl}/${id}`);
  }

  create(request: TaxRuleRequest): Observable<TaxRule> {
    return this.http.post<TaxRule>(this.baseUrl, request);
  }

  update(id: number, request: TaxRuleRequest): Observable<TaxRule> {
    return this.http.put<TaxRule>(`${this.baseUrl}/${id}`, request);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  updateStatus(id: number, request: TaxRuleStatusUpdateRequest): Observable<TaxRuleStatusUpdateResponse> {
    return this.http.patch<TaxRuleStatusUpdateResponse>(`${this.baseUrl}/${id}/status`, request);
  }

  checkNameExists(value: string, excludeId?: number): Observable<boolean> {
    let params = new HttpParams().set('value', value);
    if (excludeId != null) params = params.set('excludeId', excludeId.toString());
    return this.http.get<boolean>(`${this.baseUrl}/name-exists`, { params });
  }
}

File: backend/src/main/java/com/cms/inventory/procurement/service/TaxRuleService.java
package com.cms.inventory.procurement.service;

import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.cms.dto.ActiveStatusUpdateRequest;
import com.cms.dto.ActiveStatusUpdateResponse;
import com.cms.exception.ResourceNotFoundException;
import com.cms.inventory.procurement.dto.TaxRuleRequest;
import com.cms.inventory.procurement.dto.TaxRuleResponse;
import com.cms.inventory.procurement.model.TaxRule;
import com.cms.inventory.procurement.repository.TaxRuleRepository;

@Service
@Transactional(readOnly = true)
public class TaxRuleService {

    private final TaxRuleRepository taxRuleRepository;

    public TaxRuleService(TaxRuleRepository taxRuleRepository) {
        this.taxRuleRepository = taxRuleRepository;
    }

    @Transactional
    public TaxRuleResponse create(TaxRuleRequest request) {
        String name = requireTrimmed(request.name(), "Name is required");
        if (taxRuleRepository.existsByNameIgnoreCase(name)) {
            throw new IllegalArgumentException("A tax rule named '" + name + "' already exists");
        }

        TaxRule taxRule = new TaxRule();
        taxRule.setName(name);
        taxRule.setRatePercent(request.ratePercent());
        if (request.isActive() != null) taxRule.setIsActive(request.isActive());
        return toResponse(taxRuleRepository.save(taxRule));
    }

    public List<TaxRuleResponse> findAll(boolean activeOnly) {
        List<TaxRule> rules = activeOnly ? taxRuleRepository.findByIsActiveTrueOrderByNameAsc() : taxRuleRepository.findAllByOrderByNameAsc();
        return rules.stream().map(this::toResponse).toList();
    }

    public Page<TaxRuleResponse> findPage(String search, Pageable pageable) {
        Specification<TaxRule> spec = (root, query, cb) -> {
            if (search == null || search.isBlank()) return cb.conjunction();
            return cb.like(cb.lower(root.get("name")), "%" + search.trim().toLowerCase() + "%");
        };
        return taxRuleRepository.findAll(spec, pageable).map(this::toResponse);
    }

    public TaxRuleResponse findById(Long id) {
        return toResponse(findOrThrow(id));
    }

    @Transactional
    public TaxRuleResponse update(Long id, TaxRuleRequest request) {
        TaxRule taxRule = findOrThrow(id);
        String name = requireTrimmed(request.name(), "Name is required");
        if (taxRuleRepository.existsByNameIgnoreCaseAndIdNot(name, id)) {
            throw new IllegalArgumentException("A tax rule named '" + name + "' already exists");
        }

        taxRule.setName(name);
        taxRule.setRatePercent(request.ratePercent());
        if (request.isActive() != null) taxRule.setIsActive(request.isActive());
        return toResponse(taxRuleRepository.save(taxRule));
    }

    @Transactional
    public void delete(Long id) {
        if (!taxRuleRepository.existsById(id)) {
            throw new ResourceNotFoundException("Tax rule not found with id: " + id);
        }
        taxRuleRepository.deleteById(id);
    }

    @Transactional
    public ActiveStatusUpdateResponse updateStatus(Long id, ActiveStatusUpdateRequest request) {
        TaxRule taxRule = findOrThrow(id);
        taxRule.setIsActive(Boolean.TRUE.equals(request.isActive()));
        TaxRule saved = taxRuleRepository.save(taxRule);
        return new ActiveStatusUpdateResponse(saved.getId(), saved.getIsActive(), saved.getUpdatedAt());
    }

    public boolean nameExists(String name, Long excludeId) {
        String trimmed = name == null ? "" : name.trim();
        return excludeId != null
            ? taxRuleRepository.existsByNameIgnoreCaseAndIdNot(trimmed, excludeId)
            : taxRuleRepository.existsByNameIgnoreCase(trimmed);
    }

    TaxRule findOrThrow(Long id) {
        return taxRuleRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Tax rule not found with id: " + id));
    }

    private TaxRuleResponse toResponse(TaxRule t) {
        return new TaxRuleResponse(t.getId(), t.getName(), t.getRatePercent(), t.getIsActive(), t.getCreatedAt(), t.getUpdatedAt());
    }

    private static String trim(String s) {
        if (s == null) return null;
        String t = s.trim();
        return t.isEmpty() ? null : t;
    }

    private static String requireTrimmed(String s, String message) {
        String t = trim(s);
        if (t == null) throw new IllegalArgumentException(message);
        return t;
    }
}
